sim/core/item_swaps.go

```diff
@@ -141,91 +212,162 @@ func (swap *ItemSwap) IsEnabled() bool {
+		//will swap both on the MainHand Slot for 2H.
+		if has2H && slot == proto.ItemSlot_ItemSlotOffHand && !swap.isFuryWarrior {
+			continue
 		}
-	}
 
-	if sim.Log != nil {
-		sim.Log("Item Swap Stats: %v", newStats.FlatString())
+		if ok := swap.swapItem(slot, has2H, isReset); ok {
+			weaponSlotSwapped = slot == proto.ItemSlot_ItemSlotMainHand || slot == proto.ItemSlot_ItemSlotOffHand || slot == proto.ItemSlot_ItemSlotRanged || weaponSlotSwapped
+
+			for _, onSwap := range swap.onSwapCallbacks[slot] {
+				onSwap(sim, slot)
+			}
+		}
 	}
 
-	character.AddStatsDynamic(sim, newStats)
+	if !isReset {
+		statsToSwap := Ternary(isPrepull, swap.stats.allSlots, swap.stats.weaponSlots)
+		newStats := Ternary(swap.swapSet == proto.APLActionItemSwap_Swap1, statsToSwap.Invert(), statsToSwap)
 
-	for _, onSwap := range swap.onSwapCallbacks {
-		onSwap(sim)
+		if sim.Log != nil {
+			sim.Log("Item Swap - New Stats: %v", newStats.FlatString())
+		}
+		character.AddStatsDynamic(sim, newStats)
 	}
 
-	if character.AutoAttacks.AutoSwingMelee && meleeWeaponSwapped && sim.CurrentTime > 0 {
-		character.AutoAttacks.StopMeleeUntil(sim, sim.CurrentTime, false)
-	}
+	if !isPrepull && !isReset {
+		if weaponSlotSwapped {
+			if character.AutoAttacks.AutoSwingMelee {
+				character.AutoAttacks.StopMeleeUntil(sim, sim.CurrentTime, false)
+			}
+			if character.AutoAttacks.AutoSwingRanged {
+				character.AutoAttacks.StopRangedUntil(sim, sim.CurrentTime)
+			}
+		}
 
-	// If GCD is ready then use the GCD, otherwise we assume it's being used along side a spell.
-	if character.GCD.IsReady(sim) {
-		newGCD := sim.CurrentTime + 1500*time.Millisecond
-		character.SetGCDTimer(sim, newGCD)
+		character.ExtendGCDUntil(sim, max(character.NextGCDAt(), sim.CurrentTime+GCDDefault))
 	}
 
-	swap.swapped = !swap.swapped
+	swap.swapSet = swapSet
 }
 
-func (swap *ItemSwap) swapItem(slot proto.ItemSlot, has2H bool) (bool, stats.Stats) {
-	oldItem := swap.character.Equipment[slot]
-	newItem := swap.GetItem(slot)
+func (swap *ItemSwap) swapItem(slot proto.ItemSlot, has2H bool, isReset bool) bool {
+	oldItem := *swap.GetEquippedItemBySlot(slot)
+	var newItem *Item
+	if isReset {
+		newItem = &swap.originalEquip[slot]
+	} else {
+		newItem = swap.GetUnequippedItemBySlot(slot)
+	}
 
 	swap.character.Equipment[slot] = *newItem
-	oldItemStats := swap.getItemStats(oldItem)
-	newItemStats := swap.getItemStats(*newItem)
-	newStats := newItemStats.Subtract(oldItemStats)
 
 	//2H will swap out the offhand also.
-	if has2H && slot == proto.ItemSlot_ItemSlotMainHand {
-		_, ohStats := swap.swapItem(proto.ItemSlot_ItemSlotOffHand, has2H)
-		newStats = newStats.Add(ohStats)
+	if has2H && slot == proto.ItemSlot_ItemSlotMainHand && !swap.isFuryWarrior {
+		swap.swapItem(proto.ItemSlot_ItemSlotOffHand, has2H, isReset)
 	}
 
-	swap.unEquippedItems[slot-offset] = oldItem
+	swap.unEquippedItems[slot] = oldItem
 	swap.swapWeapon(slot)
 
-	return true, newStats
+	return true
```

Comment:
Don't need a return value here unless the refactored version might return `false`.